### PR TITLE
feature: #242 Explicit terminal and retry exceptions for cleaner logging and poison pills

### PR DESCRIPTION
**Pull request by astubbs**
- [ ] convert output of some tests to use the new exception and check logging output is as expected
- [ ] docs, including batch partial retry
- [ ] tests
- [ ] partial batch support - seperate PR
- [ ] max retries handling
- [ ] rewrite docs section: [11.2. Skipping Records](https://github.com/confluentinc/parallel-consumer/pull/291#skipping-records)
- [ ] retry dlc PR

Related:
- #195
- #65
- #242
- #34
- #196

Supersedes:
- #66

Related branches:
- retry-dlq

### CODE REVIEW

File: parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/AbstractParallelEoSStreamProcessor.java [hunk 1]
```diff
@@ -1121,7 +1122,12 @@ protected <R> List<ParallelConsumer.Tuple<ConsumerRecord<K, V>, R>> runUserFunct
             return intermediateResults;
         } catch (Exception e) {
             // handle fail
-            log.error("Exception caught in user function running stage, registering WC as failed, returning to mailbox", e);
+            String msg = "Exception caught in user function running stage, registering WC as failed, returning to mailbox";
+            if (e instanceof RetriableException) {
+                log.debug("Explicit " + RetriableException.class.getSimpleName() + " caught, logging at DEBUG only. " + msg, e);
```
[the remark this answers is not in the document]
Comment (astubbs) [reply parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/AbstractParallelEoSStreamProcessor.java:1127]: haha yeah good point :)

File: parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/AbstractParallelEoSStreamProcessor.java [hunk 2]
```diff
@@ -1121,7 +1122,12 @@ protected <R> List<ParallelConsumer.Tuple<ConsumerRecord<K, V>, R>> runUserFunct
             return intermediateResults;
         } catch (Exception e) {
             // handle fail
-            log.error("Exception caught in user function running stage, registering WC as failed, returning to mailbox", e);
+            String msg = "Exception caught in user function running stage, registering WC as failed, returning to mailbox";
+            if (e instanceof RetriableException) {
```
[the remark this answers is not in the document]
Comment (astubbs) [reply parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/AbstractParallelEoSStreamProcessor.java:1126]: It's mostly for the common failure handling logic and message reuse.
The throw wouldn't be caught here though.

The alternative would look like:

        } catch (RetriableException e) {
            String msg = "Exception caught in user function running stage, registering WC as failed, returning to mailbox";
            log.debug("Explicit " + RetriableException.class.getSimpleName() + " caught, logging at DEBUG only. " + msg, e);
            handleFailure(workContainerBatch, e);
            throw e; // trow again to make the future failed
        } catch (Exception e) {
            log.error("Exception caught in user function running stage, registering WC as failed, returning to mailbox", e);
            handleFailure(workContainerBatch, e);
            throw e; // trow again to make the future failed
        }
    }

    private void handleFailure(final List<? extends WorkContainer<K, V>> workContainerBatch, final Exception e) {
        for (var wc : workContainerBatch) {
            wc.onUserFunctionFailure(e);
            addToMailbox(wc); // always add on error
        }
    }

[the remark this answers is not in the document]
Comment (astubbs) [reply parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/AbstractParallelEoSStreamProcessor.java:1126]: let me know what you think :)

**Comment by astubbs** [comment]
#268 ?

File: parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/AbstractParallelEoSStreamProcessor.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (ifeeneyconfluent) [reply parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/AbstractParallelEoSStreamProcessor.java]: Ah yes, the log level param does look a nicer.  Honestly though, I prefer the explicit handler for RetriableException, but many ways to skin a cat! ;-)

**Review by ifeeneyconfluent — APPROVED** [review]
FWIW, I'm happy to approve.

**Review by rkolesnev — CHANGES_REQUESTED** [review]

File: parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/AbstractParallelEoSStreamProcessor.java [hunk 3]
```diff
@@ -1099,37 +1101,101 @@ protected <R> List<ParallelConsumer.Tuple<ConsumerRecord<K, V>, R>> runUserFunct
...
+
+                // return empty result to cause system to skip as if it succeeded
+                return new ArrayList<>();
+            } else if (reaction == SHUTDOWN) {
+                log.error("Shutting down upon terminal failure in user function due to {} setting in {} - triggering context: {}",
+                        ParallelConsumerOptions.TerminalFailureReaction.class.getSimpleName(),
+                        ParallelConsumerOptions.class.getSimpleName(),
+                        context);
+
+                closeDontDrainFirst();
```
[the remark this answers is not in the document]
Comment (astubbs) [reply parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/AbstractParallelEoSStreamProcessor.java]: yeah, it is graceful - `closeDontDrainFirst` will commit everything that's done first - drain isn't graceful, it means process everything in buffers.
Inflight - good point, not 100% will need to double check.

File: parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/AbstractParallelEoSStreamProcessor.java [hunk 4]
```diff
@@ -1099,37 +1101,101 @@ protected <R> List<ParallelConsumer.Tuple<ConsumerRecord<K, V>, R>> runUserFunct
...
             }
-            throw e; // trow again to make the future failed
+        } catch (Exception e) {
+            log.error("Unknown internal error handling user function dispatch, terminating");
+
+            closeDontDrainFirst();
+
+            // throw again to make the future failed
+            throw e;
         }
```
[the remark this answers is not in the document]
Comment (astubbs) [reply parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/AbstractParallelEoSStreamProcessor.java]: current = retry :)

> wouldn't this just kill PC and not retry the message?

yeah there was a bug - check latest version?

**Comment by astubbs** [comment]
Consider splitting the partial batch failure support into a seperate task.

File: parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/PCUserException.java [hunk 5]
```diff
@@ -0,0 +1,22 @@
+package io.confluent.parallelconsumer;
+
+/*-
+ * Copyright (C) 2020-2022 Confluent, Inc.
+ */
+
+/**
+ * todo
```
Comment (astubbs) [inline parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/PCUserException.java:8]: docs

File: parallel-consumer-core/src/test/java/io/confluent/parallelconsumer/internal/UserFunctionRunnerTest.java [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+import static io.confluent.parallelconsumer.ParallelConsumerOptions.TerminalFailureReaction.SHUTDOWN;
+import static io.confluent.parallelconsumer.ParallelConsumerOptions.TerminalFailureReaction.SKIP;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+/**
+ * @see UserFunctionRunner
+ */
+class UserFunctionRunnerTest {
```
Comment (astubbs) [inline parallel-consumer-core/src/test/java/io/confluent/parallelconsumer/internal/UserFunctionRunnerTest.java:26]: finish test

File: parallel-consumer-core/src/test/resources/logback-test.xml [hunk 7]
```diff
@@ -27,8 +27,8 @@
     </root>
 
     <!-- primary -->
-    <logger name="io.confluent.parallelconsumer" level="info"/>
-<!--        <logger name="io.confluent.parallelconsumer" level="debug"/>-->
+    <logger name="io.confluent.parallelconsumer" level="debug"/>
```
Comment (astubbs) [inline parallel-consumer-core/src/test/resources/logback-test.xml:30]: reset

**Review by astubbs — COMMENTED** [review]
...

File: parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/UserFunctionRunner.java [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+            // throw again to make the future failed
+            throw e;
+        } catch (Exception e) {
+            log.error("Unknown internal error handling user function dispatch, terminating");
+
+            pc.closeDontDrainFirst();
+
+            // throw again to make the future failed
+            throw e;
+        }
```
[the remark this answers is not in the document]
Comment (astubbs) [reply parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/UserFunctionRunner.java:70]: This needs to be optional behaviour - I thought I had implemented that in this PR, but might be somewhere else. Either way, it should do like KS does - and have this be configurable.
I'll also probably copy their default behaviour.

**Comment by eddyv** [comment]
Closing - Stale.

**Comment by nizarsalhaji94** [comment]
Hi @astubbs 

Is a released date is planned ? 
I’m interested to use SHUTDOWN Terminal Failure Reatction.
In fact, we need to shutdown (or stop or kill) the PC after max retry. 

Thanks

**Comment by astubbs** [comment]
@nizarsalhaji94 hi sorry I am no longer involved with this project since February 2023. Try asking @eddyv

**Comment by nizarsalhaji94** [comment]
Hi @eddyv 

Is a released date is planned ?
I’m interested to use SHUTDOWN Terminal Failure Reatction.
In fact, we need to shutdown (or stop or kill) the PC after max retry.

Thanks

**Comment by rkolesnev** [comment]
@nizarsalhaji94
We are planning to implement this feature at some point, but no concrete dates / roadmap for this yet.
As a work around - i have suggested to do the stop/kill from user function approach - in this issue https://github.com/confluentinc/parallel-consumer/issues/718#issuecomment-1945771005

Just in general - i haven't seen your comment here until now - in the future - it would be more visible if you raise it as a question / issue with link to the PR if its relevant.

**Pull request closed without merging**